File: 백준/Silver/9012. 괄호/괄호.py
import sys
input = sys.stdin.readline
t = int(input())
for _ in range(t):
    str = ''.join(input().rstrip().split('()'))
    while True:
        new_str = ''.join(str.split('()'))
        if new_str == str:
            break
        else:
            str = new_str
    if str == '':
        print('YES')
    else:
        print('NO')

File: 백준/Silver/9012. 괄호/괄호.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Stack;

public class Main {

	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		Stack<Integer> stack = new Stack<>();
		String input;
		int t = Integer.parseInt(br.readLine());
		
		for (int tc=0; tc<t; tc++) {
			stack.removeAll(stack);
			input = br.readLine();
			for (int i=0; i<input.length(); i++) {
				if (input.charAt(i) == '(') stack.push(0);
				else {
					if (stack.size() == 0) {
						stack.push(0);
						break;
					}
					stack.pop();
				}
			}
			System.out.println(stack.size() == 0 ? "YES" : "NO");
		}

	}
}
